File: java/datastructure/linkedlist/leetcode/MiddleNode.java
package datastructure.linkedlist.leetcode;

import java.util.ArrayList;
import java.util.List;

/**
 * @author roseduan[roseduan520@gmail.com]
 * @time 2019/12/6 0:21
 * @description 找到链表的中间节点
 */
public class MiddleNode {

    /**
     * 使用一个数组
     */
    public ListNode findMiddleNode1(ListNode head) {
        List<ListNode> list = new ArrayList<>();
        while (head != null) {
            list.add(head);
            head = head.next;
        }
        return list.get(list.size() / 2);
    }
    
    /**
     * 计数法，两次遍历
     */
    public ListNode findMiddleNode2(ListNode head) {
        int count = 0;
        ListNode p = head;
        while (p != null) {
            count++;
            p = p.next;
        }

        count /= 2;
        p = head;
        while (count-- > 0) {
            p = p.next;
        }
        return p;
    }

    /**
     * 双指针法
     */
    public ListNode findMiddleNode3(ListNode head) {
        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }

    private static class ListNode {
        int val;
        ListNode next;

        ListNode(int val) {
            this.val = val;
            this.next = null;
        }
    }
}

File: go/linkedlist/leetcode/middle_node.go
package main

// 链表的中间节点

// 使用一个数组存储
func middleNode1(head *ListNode) *ListNode {
	var arr []*ListNode
	for head != nil {
		arr = append(arr, head)
		head = head.Next
	}
	return arr[len(arr)/2]
}

// 计数法，两次遍历
func middleNode2(head *ListNode) *ListNode {
	p, count := head, 0
	for p != nil {
		p, count = p.Next, count+1
	}

	p, count = head, count/2
	for count > 0 {
		p, count = p.Next, count-1
	}
	return p
}

// 快慢指针法
func middleNode3(head *ListNode) *ListNode {
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow, fast = slow.Next, fast.Next.Next
	}
	return slow
}
